utils/yoke/src/yoke.rs

```diff
@@ -0,0 +1,342 @@
+    /// #         let ret = mem::transmute_copy(&from);
+    /// #         mem::forget(from);
+    /// #         ret
+    /// #     }
+    /// #
+    /// #     fn with_mut<F>(&'a mut self, f: F)
+    /// #     where
+    /// #         F: 'static + FnOnce(&'a mut Self::Output),
+    /// #     {
+    /// #         unsafe { f(mem::transmute(self)) }
+    /// #     }
+    /// # }
+    /// ```
+    pub fn with_mut<'a, F>(&'a mut self, f: F)
+    where
+        F: 'static + for<'b> FnOnce(&'b mut <Y as Yokeable<'a>>::Output),
+    {
+        self.yokeable.with_mut(f)
+    }
+}
+
+impl<Y: for<'a> Yokeable<'a>, C: StableDeref> Yoke<Y, Option<C>> {
+    /// Construct a new [`Yoke`] from static data. There will be no
+    /// references to `cart` here since [`Yokeable`]s are `'static`,
+    /// this is good for e.g. constructing fully owned
+    /// [`Yoke`]s with no internal borrowing.
+    ///
+    /// This can be paired with [`Yoke::attach_to_option_cart()`] to mix owned
+    /// and borrowed data.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use yoke::Yoke;
+    /// # use std::borrow::Cow;
+    /// # use std::rc::Rc;
+    ///
+    /// let owned: Cow<str> = "hello".to_owned().into();
+    /// // this yoke can be intermingled with actually-borrowed Yokes
+    /// let yoke: Yoke<Cow<str>, Option<Rc<[u8]>>> = Yoke::new_owned(owned);
+    ///
+    /// assert_eq!(yoke.get(), "hello");
+    /// ```
+    pub fn new_owned(yokeable: Y) -> Self {
+        Self {
+            yokeable,
+            cart: None,
+        }
+    }
+
+    /// Similar to [`Yoke::attach_to_cart()`], except it constructs a `Yoke<Y, Option<C>>`
+    /// instead, where the cart is `Some(..)`.
+    ///
+    /// This allows mixing [`Yoke`]s constructed from owned and borrowed data, when
+    /// paired with [`Yoke::new_owned()`].
+    ///
+    /// This method is currently unusable due to a [compiler bug](https://github.com/rust-lang/rust/issues/84937),
+    /// use [`Yoke::attach_to_option_cart_badly()`] instead
+    pub fn attach_to_option_cart<F>(cart: C, f: F) -> Self
+    where
+        F: for<'de> FnOnce(&'de <C as Deref>::Target) -> <Y as Yokeable<'de>>::Output,
+    {
+        let deserialized = f(cart.deref());
+        Self {
+            yokeable: unsafe { Y::make(deserialized) },
+            cart: Some(cart),
+        }
+    }
+    /// Temporary version of [`Yoke::attach_to_option_cart()`]
+    /// that doesn't hit https://github.com/rust-lang/rust/issues/84937
+    ///
+    /// See its docs for more details
+    pub fn attach_to_option_cart_badly(
+        cart: C,
+        f: for<'de> fn(&'de <C as Deref>::Target) -> <Y as Yokeable<'de>>::Output,
+    ) -> Self {
+        let deserialized = f(cart.deref());
+        Self {
+            yokeable: unsafe { Y::make(deserialized) },
+            cart: Some(cart),
+        }
+    }
+}
+
+// clone impls only work for reference counted objects, otherwise you should be
```

Comment:
Suggestion: use `CloneStableDeref` here to make this a bit more general.  It works for Rc, Arc, and `&'a T`.

Reply:
Hmm, so this works, but then you can no longer have the impl on `Option<Rc<T>>` (since `stable_deref_trait` could potentially introduce a `impl CloneStableDeref for Option<T>` -- it wouldn't make sense but the compiler doesn't know that)

Reply:
Okay, can you at least add the missing impls?  Consider using `macro_rules` to make it shorter.